src/ast/document-descriptor.ts

```diff
@@ -38,3 +46,206 @@ export class DocumentDescriptor implements Descriptor {
     return this.document.url;
   }
 }
+
+export interface Feature {
+  kinds: Iterable<string>;
+  identifiers?: Iterable<string>;
+  // sourceLocation: SourceLocation;
+}
+
+
+export interface Resolvable extends ScannedFeature {
+  resolve(document: Document): Feature;
+}
+function isResolvable(x: any): x is Resolvable {
+  return x.resolve && typeof x.resolve === 'function';
+}
+
+export class Document implements Feature {
+  url: string;
+  parsedDocument: ParsedDocument<any, any>;
+  isInline: boolean;
+
+  kinds: Iterable<string>;
+  identifiers: Set<string>;
+
+  private _rootDocument: Document;
```

Comment:
I think it's better to match semantics first and warn second. It is very common to depend on implicit deps.